java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,599 @@
+
+    String response = Sage.readLineBytes(inStream);
+    if (!"OK".equals(response))
+    {
+      // Don't leave the connection open. We can't read anything if we are seeing this error anyway.
+      disconnect();
+      throw new IOException("Error opening remote file of:" + response);
+    }
+  }
+
+  private void disconnect()
+  {
+    if (socket != null)
+    {
+      try
+      {
+        outStream.write("QUIT\r\n".getBytes(Sage.BYTE_CHARSET));
+        outStream.flush();
+      }
+      catch (IOException e) {}
+
+      try
+      {
+        socket.close();
+      }
+      catch (IOException e) {}
+      socket = null;
+    }
+    if (outStream != null)
+    {
+      try
+      {
+        outStream.close();
+      }
+      catch (IOException e) {}
+      outStream = null;
+    }
+    if (inStream != null)
+    {
+      try
+      {
+        inStream.close();
+      }
+      catch (IOException e) {}
+      inStream = null;
+    }
+  }
+
+  private synchronized void reconnect() throws IOException
+  {
+    disconnect();
+    connect();
+  }
+
+  private long getMaxRead(long position, long count) throws IOException
+  {
+    // We can't read more than what is in the file. The first part checks a cached variable, the
+    // second part checks the file directly. Unless the stream really lingers trying to get a
+    // massive read towards the end, this should be a very minimal load on the media server.
+    count = Math.min(
+        maxRemoteSize == 0 || (position + count >= maxRemoteSize && activeFile) ?
+            reallyGetLength() - position : maxRemoteSize - position,
+        count);
+
+    // If we are trying to read beyond the end of the file, don't read anything.
+    return count;
+  }
+
+  @Override
+  public synchronized int read() throws IOException
+  {
+    int returnValue;
+
+    if (getMaxRead(remoteOffset, 1) == 0)
+    {
+      // End of the file.
+      return -1;
+    }
+
+    try
+    {
+      outStream.write(("READ " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      returnValue = inStream.read();
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("READ " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      returnValue = inStream.read();
+    }
+
+    remoteOffset += 1;
+    return returnValue & 0xFF;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException
+  {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int off, int len) throws IOException
+  {
+    if (len == 0)
+      return len;
+
+    // This cannot return anything greater than len.
+    len = (int)getMaxRead(remoteOffset, len);
+
+    // End of the file.
+    if (len == 0)
+      return -1;
+
+    try
+    {
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      inStream.readFully(b, off, len);
+      outStream.flush();
```

Comment:
Fixed it and I noticed that I did something less than efficient with readFully, so I fixed that too.